File: net/crates/net/bindings/go/net/capability_aggregation_test.go
// JSON-encoder pins for the Phase 6c capability-aggregation surface.
//
// The Rust core's `serde_json::to_string` produces specific byte
// sequences pinned by `serde_shapes_match_cross_binding_wire_format`
// in `capability_aggregation.rs`. The Go encoder in this file must
// produce the same JSON for the C ABI to deserialize correctly.
// This file pins the Go side; both sides move together when the
// wire shape changes.
//
// Tests are pure marshaling — no cgo round-trip, so they run as
// part of plain `go test ./...` without a built `libnet`.
package net

import (
	"encoding/json"
	"testing"
)

// ──────────────────────────────────────────────────────────────────
// TagMatcher
// ──────────────────────────────────────────────────────────────────

func TestMatchExact(t *testing.T) {
	b, err := json.Marshal(MatchExact("software.python=3.11"))
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	if got, want := string(b), `{"kind":"exact","value":"software.python=3.11"}`; got != want {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestMatchPrefix(t *testing.T) {
	b, _ := json.Marshal(MatchPrefix("hardware.gpu"))
	if got, want := string(b), `{"kind":"prefix","value":"hardware.gpu"}`; got != want {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestMatchAxis(t *testing.T) {
	b, _ := json.Marshal(MatchAxis(AxisHardware))
	if got, want := string(b), `{"kind":"axis","axis":"hardware"}`; got != want {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestMatchAxisKey(t *testing.T) {
	b, _ := json.Marshal(MatchAxisKey(AxisHardware, "gpu.count"))
	if got, want := string(b), `{"kind":"axis_key","axis":"hardware","key":"gpu.count"}`; got != want {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestMatchRegex(t *testing.T) {
	b, _ := json.Marshal(MatchRegex("^a$"))
	if got, want := string(b), `{"kind":"regex","pattern":"^a$"}`; got != want {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestMatchVersionRangeWithMinOnly(t *testing.T) {
	min := "3.10.0"
	b, _ := json.Marshal(MatchVersionRange("software.python", &min, nil))
	if got, want := string(b),
		`{"kind":"version_range","axis_key":"software.python","min":"3.10.0"}`; got != want {
		t.Errorf("got %s, want %s", got, want)
	}
}

// ──────────────────────────────────────────────────────────────────
// GroupBy
// ──────────────────────────────────────────────────────────────────

func TestGroupByVariants(t *testing.T) {
	cases := []struct {
		name string
		gb   GroupBy
		want string
	}{
		{"Class", GroupByClass(), `{"kind":"class"}`},
		{"State", GroupByState(), `{"kind":"state"}`},
		{"Region", GroupByRegion(), `{"kind":"region"}`},
		{"Publisher", GroupByPublisher(), `{"kind":"publisher"}`},
		{
			"TagStem",
			GroupByTagStem("hardware.gpu"),
			`{"kind":"tag_stem","prefix":"hardware.gpu"}`,
		},
		{
			"TagValue",
			GroupByTagValue(AxisSoftware, "python"),
			`{"kind":"tag_value","axis":"software","key":"python"}`,
		},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			b, _ := json.Marshal(c.gb)
			if got := string(b); got != c.want {
				t.Errorf("got %s, want %s", got, c.want)
			}
		})
	}
}

// ──────────────────────────────────────────────────────────────────
// Aggregation
// ──────────────────────────────────────────────────────────────────

func TestAggregationVariants(t *testing.T) {
	cases := []struct {
		name string
		agg  Aggregation
		want string
	}{
		{"Count", AggCount(), `{"kind":"count"}`},
		{
			"DistinctPublishers",
			AggDistinctPublishers(),
			`{"kind":"distinct_publishers"}`,
		},
		{
			"DistinctValues",
			AggDistinctValues(AxisSoftware, "python"),
			`{"kind":"distinct_values","axis":"software","key":"python"}`,
		},
		{
			"SumNumericTag",
			AggSumNumericTag("hardware.gpu.count"),
			`{"kind":"sum_numeric_tag","axis_key":"hardware.gpu.count"}`,
		},
		{
			"MinNumericTag",
			AggMinNumericTag("hardware.gpu.count"),
			`{"kind":"min_numeric_tag","axis_key":"hardware.gpu.count"}`,
		},
		{
			"MaxNumericTag",
			AggMaxNumericTag("hardware.gpu.count"),
			`{"kind":"max_numeric_tag","axis_key":"hardware.gpu.count"}`,
		},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			b, _ := json.Marshal(c.agg)
			if got := string(b); got != c.want {
				t.Errorf("got %s, want %s", got, c.want)
			}
		})
	}
}

// ──────────────────────────────────────────────────────────────────
// CapacityQuery — round-trip + nil-field handling
// ──────────────────────────────────────────────────────────────────

func TestCapacityQueryRoundTripsRustWireShape(t *testing.T) {
	matcher := MatchPrefix("hardware.gpu")
	maxRtt := uint32(50)
	q := CapacityQuery{
		Matcher:    &matcher,
		GroupBy:    GroupByTagStem("hardware.gpu"),
		MaxRTTMs:   &maxRtt,
		SumAxisKey: "hardware.gpu.count",
		Limit:      5,
	}
	b, err := json.Marshal(q)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var got map[string]any
	if err := json.Unmarshal(b, &got); err != nil {
		t.Fatalf("re-parse: %v", err)
	}
	if got["limit"].(float64) != 5 {
		t.Errorf("limit: got %v", got["limit"])
	}
	if got["max_rtt_ms"].(float64) != 50 {
		t.Errorf("max_rtt_ms: got %v", got["max_rtt_ms"])
	}
	if got["sum_axis_key"].(string) != "hardware.gpu.count" {
		t.Errorf("sum_axis_key: got %v", got["sum_axis_key"])
	}
	m := got["matcher"].(map[string]any)
	if m["kind"].(string) != "prefix" || m["value"].(string) != "hardware.gpu" {
		t.Errorf("matcher: got %v", m)
	}
	g := got["group_by"].(map[string]any)
	if g["kind"].(string) != "tag_stem" || g["prefix"].(string) != "hardware.gpu" {
		t.Errorf("group_by: got %v", g)
	}
}

func TestCapacityQueryOmitsAbsentOptionalFields(t *testing.T) {
	q := CapacityQuery{GroupBy: GroupByRegion(), Limit: 0}
	b, _ := json.Marshal(q)
	var got map[string]any
	if err := json.Unmarshal(b, &got); err != nil {
		t.Fatalf("re-parse: %v", err)
	}
	// Optional fields are omitted (not nulled) per Go's omitempty
	// convention. The Rust core's CapacityQuery has serde defaults
	// for these — `Option<TagMatcher>` defaults None, `max_rtt_ms`
	// defaults None — so omission is equivalent to null at the
	// Rust boundary.
	if _, ok := got["matcher"]; ok {
		t.Errorf("matcher should be omitted; got %v", got["matcher"])
	}
	if _, ok := got["max_rtt_ms"]; ok {
		t.Errorf("max_rtt_ms should be omitted")
	}
	if got["limit"].(float64) != 0 {
		t.Errorf("limit: got %v", got["limit"])
	}
}

File: net/crates/net/sdk-ts/test/capability_aggregation.test.ts
// JSON-encoder pins for the Phase 6c capability-aggregation surface.
//
// The Rust core's `serde_json::to_string(&TagMatcher::...)` etc.
// produces specific byte sequences pinned by
// `serde_shapes_match_cross_binding_wire_format` in
// `capability_aggregation.rs`. The TS encoder in
// `src/capability-aggregation.ts` must produce the same JSON for the
// napi boundary to deserialize correctly. This file pins the TS
// side; if either side moves, both move together.

import { describe, expect, it } from 'vitest';

import {
  aggregationToJson,
  capacityQueryToJson,
  groupByToJson,
  tagMatcherToJson,
} from '../src/capability-aggregation';

describe('tagMatcherToJson', () => {
  it('encodes Exact as { kind, value }', () => {
    expect(
      tagMatcherToJson({ kind: 'exact', value: 'software.python=3.11' }),
    ).toBe('{"kind":"exact","value":"software.python=3.11"}');
  });

  it('encodes Prefix as { kind, value }', () => {
    expect(tagMatcherToJson({ kind: 'prefix', value: 'hardware.gpu' })).toBe(
      '{"kind":"prefix","value":"hardware.gpu"}',
    );
  });

  it('encodes Axis as { kind, axis }', () => {
    expect(tagMatcherToJson({ kind: 'axis', axis: 'hardware' })).toBe(
      '{"kind":"axis","axis":"hardware"}',
    );
  });

  it('encodes AxisKey as { kind, axis, key }', () => {
    expect(
      tagMatcherToJson({ kind: 'axisKey', axis: 'hardware', key: 'gpu.count' }),
    ).toBe('{"kind":"axis_key","axis":"hardware","key":"gpu.count"}');
  });

  it('encodes Regex as { kind, pattern }', () => {
    expect(tagMatcherToJson({ kind: 'regex', pattern: '^a$' })).toBe(
      '{"kind":"regex","pattern":"^a$"}',
    );
  });

  it('encodes VersionRange with bounds nulled when omitted', () => {
    expect(
      tagMatcherToJson({
        kind: 'versionRange',
        axisKey: 'software.python',
        min: '3.10.0',
      }),
    ).toBe(
      '{"kind":"version_range","axis_key":"software.python","min":"3.10.0","max":null}',
    );
  });
});

describe('groupByToJson', () => {
  it('encodes Class', () => {
    expect(groupByToJson({ kind: 'class' })).toBe('{"kind":"class"}');
  });

  it('encodes State', () => {
    expect(groupByToJson({ kind: 'state' })).toBe('{"kind":"state"}');
  });

  it('encodes Region', () => {
    expect(groupByToJson({ kind: 'region' })).toBe('{"kind":"region"}');
  });

  it('encodes Publisher', () => {
    expect(groupByToJson({ kind: 'publisher' })).toBe('{"kind":"publisher"}');
  });

  it('encodes TagStem as { kind, prefix }', () => {
    expect(groupByToJson({ kind: 'tagStem', prefix: 'hardware.gpu' })).toBe(
      '{"kind":"tag_stem","prefix":"hardware.gpu"}',
    );
  });

  it('encodes TagValue as { kind, axis, key }', () => {
    expect(
      groupByToJson({ kind: 'tagValue', axis: 'software', key: 'python' }),
    ).toBe('{"kind":"tag_value","axis":"software","key":"python"}');
  });
});

describe('aggregationToJson', () => {
  it('encodes Count', () => {
    expect(aggregationToJson({ kind: 'count' })).toBe('{"kind":"count"}');
  });

  it('encodes DistinctPublishers', () => {
    expect(aggregationToJson({ kind: 'distinctPublishers' })).toBe(
      '{"kind":"distinct_publishers"}',
    );
  });

  it('encodes DistinctValues as { kind, axis, key }', () => {
    expect(
      aggregationToJson({
        kind: 'distinctValues',
        axis: 'software',
        key: 'python',
      }),
    ).toBe('{"kind":"distinct_values","axis":"software","key":"python"}');
  });

  it('encodes SumNumericTag as { kind, axis_key }', () => {
    expect(
      aggregationToJson({
        kind: 'sumNumericTag',
        axisKey: 'hardware.gpu.count',
      }),
    ).toBe('{"kind":"sum_numeric_tag","axis_key":"hardware.gpu.count"}');
  });

  it('encodes MinNumericTag / MaxNumericTag with the right discriminant', () => {
    expect(
      aggregationToJson({
        kind: 'minNumericTag',
        axisKey: 'hardware.gpu.count',
      }),
    ).toBe('{"kind":"min_numeric_tag","axis_key":"hardware.gpu.count"}');
    expect(
      aggregationToJson({
        kind: 'maxNumericTag',
        axisKey: 'hardware.gpu.count',
      }),
    ).toBe('{"kind":"max_numeric_tag","axis_key":"hardware.gpu.count"}');
  });
});

describe('capacityQueryToJson', () => {
  it('round-trips through Rust core wire shape', () => {
    const json = capacityQueryToJson({
      matcher: { kind: 'prefix', value: 'hardware.gpu' },
      groupBy: { kind: 'tagStem', prefix: 'hardware.gpu' },
      maxRttMs: 50,
      sumAxisKey: 'hardware.gpu.count',
      limit: 5,
    });
    expect(JSON.parse(json)).toEqual({
      matcher: { kind: 'prefix', value: 'hardware.gpu' },
      group_by: { kind: 'tag_stem', prefix: 'hardware.gpu' },
      max_rtt_ms: 50,
      sum_axis_key: 'hardware.gpu.count',
      limit: 5,
    });
  });

  it('nulls absent fields', () => {
    const json = capacityQueryToJson({
      groupBy: { kind: 'region' },
      limit: 0,
    });
    expect(JSON.parse(json)).toEqual({
      matcher: null,
      group_by: { kind: 'region' },
      max_rtt_ms: null,
      sum_axis_key: null,
      limit: 0,
    });
  });
});
